README.md
# Event Dispatcher

[![Build Status](https://travis-ci.org/FerMod/EventDispatcher.svg?branch=master)](https://travis-ci.org/FerMod/EventDispatcher) [![Sonar Quality Gate](https://img.shields.io/sonar/https/sonarcloud.io/com.fermod:event-dispatcher/quality_gate.svg)](https://sonarcloud.io/dashboard?id=com.fermod%3Aevent-dispatcher) [![Sonar Coverage](https://img.shields.io/sonar/https/sonarcloud.io/com.fermod:event-dispatcher/coverage.svg)](https://sonarcloud.io/dashboard?id=com.fermod%3Aevent-dispatcher) [![GitHub release](https://img.shields.io/github/release-pre/FerMod/EventDispatcher.svg)](https://github.com/FerMod/EventDispatcher/releases)

Event Dispatcher is simple library that offers basic event management.

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See the section [Using the Library](#using-the-library) for notes on how to use the library and how to add custom events.

### Prerequisites

This project requires to have **JDK 1.8**, and **Maven** which will manage all the remaining dependencies. **Doxygen** is only required when generating documentation from the annotated sources.

### Installing

To install the library is required to follow a few steps:

1. Go to the project [releases](https://github.com/FerMod/EventDispatcher/releases/latest).
2. Download the `jar` and place it inside your project.
3. Run the following Maven command replacing `PATH_TO_JAR` with the location of the downloaded `jar`:

    ```bash
    mvn org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install-file -Dfile=PATH_TO_JAR
    ```

4. Add to your `pom.xml` file the library dependency, it should look similar to the example below. Replace `VERSION` with the current version. If the version is `v0.1.5-beta` the value that should be replaced with is `0.1.5-beta`.

    ```xml
    <dependency>
        <groupId>com.fermod</groupId>
        <artifactId>event-dispatcher</artifactId>
        <version>VERSION</version>
    </dependency>
    ```

## Using the Library

The events can be added using the class `EventPublisher`, this class needs to be extended in order to use the methods that allow to register and fire the events. The class `ObservedValue` shows a simple implementation and is a class ready to use. 

The class `ObservedValue` extends `EventPublisher` and uses the methods declared in the interface `ValueChangeListener`.

```java
public class ObservedValue<T> extends EventPublisher<ValueChangeListener<T>> implements Serializable {
    (...)
}
```

![Diagram](src/main/resources/EventDisp_Diagram.png)

When the events need to be fired the class only need to notify the listeners. For example, in the `set` method is notified to the listeners that the value have changed, and the old and new values are passed as parameters, that will obtain the registered listeners.

```java
public void set(T value) {

    T oldValue = this.value;
    this.value = value;

    // Notify the list of registered listeners
    this.notifyListeners(listener -> listener.onValueChanged(oldValue, value));

}
```

To use the `ObservedValue` class first, it needs to be imported.

```java
import com.fermod.observer.ObservedValue;
```

When registering the listeners, it can be done inline or by passing a reference to the method.

```java
ObservedValue<Integer> observedValue = new ObservedValue<>(0);

// Inline listener registration using lambda
observedValue.registerListener((oldVAlue, newValue) -> {
    System.out.println("Value changed, executed lamda inline code. Old value: " + oldVAlue + " New value: " + newValue);
});

// Equivalent inline listener registration of the lambda
observedValue.registerListener(new ValueChangeListener<Integer>() {
    @Override
    public void onValueChanged(Integer oldValue, Integer value) {
        System.out.println("Value changed, executed inline code. Old value: " + oldValue + " New value: " + value);
    }
});
```

A method reference can be used to register as listener.

```java
// Register a mothod using method references
ObservedValue<Integer> observedValue = new ObservedValue<>(0);
observedValue.registerListener(this::printValueChange);
```

The method that is referenced is the following:

```java
public void printValueChange(int oldValue, int newValue) {
    System.out.println("Value changed, executed printValueChange method. Old value: " + oldValue + " New value: " + newValue);
}
```

All the last methods would be called when using the `set` method, and each of them would execute its content.

```java
import com.fermod.event.ValueChangeListener;
import com.fermod.observer.ObservedValue;

public class Examples {

    public static void main(String[] args) {

        ObservedValue<Integer> observedValue = new ObservedValue<>(0);

        // Inline listener registration using lambda
        observedValue.registerListener((oldVAlue, newValue) -> {
            System.out.println("Executed lambda inline code. Old value: " + oldVAlue + " New value: " + newValue);
        });

        // Equivalent inline listener registration of the lambda
        observedValue.registerListener(new ValueChangeListener<Integer>() {
            @Override
            public void onValueChanged(Integer oldValue, Integer value) {
                System.out.println("Executed inline code. Old value: " + oldValue + " New value: " + value);
            }
        });

        // Register a mothod using method references
        observedValue.registerListener(Examples::printValueChange);

        System.out.println("Change value to 10");
        observedValue.set(10);

    }

    public static void printValueChange(int oldValue, int newValue) {
        System.out.println("Executed printValueChange method. Old value: " + oldValue + " New value: " + newValue);
    }

}
```

The console output generated by this example would be:

```text
Change value to 10
Executed lambda inline code. Old value: 0 New value: 10
Executed inline code. Old value: 0 New value: 10
Executed printValueChange method. Old value: 0 New value: 10
```

The full example class can be found [here](src/test/java/com/fermod/example/Examples.java).

## Running the tests

This project uses JUnit5 tests and they are configured to run automatically in the `test` life cycle of Maven. They can also be run with the command `mvn test`. 
Along with the code unit test, the code coverage is also measured. The test and code coverage reports generate in the `site` life cycle of Maven, or when using the `mvn site` command.

This repository is configured to use Travis CI, that offers continuous integration, and it automatically builds and runs tests when pushing to the repository.

## Built With

- [Maven](https://maven.apache.org/) - Dependency management
- [JUnit5](https://junit.org/junit5/) - Unit testing framework for Java
- [TravisCI](https://travis-ci.org/) - Continous integration service
- [Doxygen](http://www.doxygen.nl/) - The documentation generation from annotated sources tool

## Contributing

Please read [CONTRIBUTING](CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/FerMod/EventDispatcher/tags).

## Authors

- [FerMod](https://github.com/FerMod) **(Ferran Tudela)** - *Initial work*

See also the list of [contributors](https://github.com/FerMod/EventDispatcher/contributors) who participated in this project.

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.

## Changelog

To know more about the changes between versions, you can go to the [CHANGELOG](CHANGELOG.md) file.

---

The general project structure tree is the following:

```tex
.
├── docs
│   └── (...)
├── doxygen
│   ├── html -> ../docs
│   ├── images
│   │   └── (...)
│   ├── customdoxygen.css
│   ├── Doxyfile
│   ├── DoxygenLayout.xml
│   ├── footer.html
│   └── header.html
├── src
│   ├── main
│   │   ├── java
│   │   │   └── com
│   │   │       └── fermod
│   │   │           └── (...)
│   │   └── resources
│   │       └── log4j2.xml
│   └── test
│       ├── java
│       │   └── com
│       │       └── fermod
│       │           └── (...)
│       └── resources
├── target
│   ├── classes
│   │   ├── com
│   │   │   └── fermod
│   │   │       ├── event
│   │   │       │   └── (...)
│   │   │       └── observer
│   │   │           └── (...)
│   │   ├── META-INF
│   │   │   ├── maven
│   │   │   │   └── com.fermod
│   │   │   │       └── event-dispatcher
│   │   │   │           ├── pom.properties
│   │   │   │           └── pom.xml
│   │   │   └── MANIFEST.MF
│   │   └── log4j2.xml
│   ├── generated-sources
│   │   └── annotations
│   ├── generated-test-sources
│   │   └── test-annotations
│   ├── logs
│   │   └── (...)
│   ├── maven-status
│   │   └── maven-compiler-plugin
│   │       ├── compile
│   │       │   └── default-compile
│   │       │       ├── createdFiles.lst
│   │       │       └── inputFiles.lst
│   │       └── testCompile
│   │           └── default-testCompile
│   │               ├── createdFiles.lst
│   │               └── inputFiles.lst
│   ├── surefire-reports
│   │   ├── com.fermod.ObservableValueTest.txt
│   │   └── TEST-com.fermod.ObservableValueTest.xml
│   └── test-classes
│       └── com
│           └── fermod
│               └── (...)
├── LICENSE
├── pom.xml
└── README.md
```


CHANGELOG.md
# Changelog

## [Unreleased][unreleased]

These changes have not been released yet.

- Fixed automatic release name generation based on the tag name

## [0.1.6-beta.1][v0.1.6-beta.1] (2019-05-22)

- Improved docs
- Refactored unit tests

## [0.1.6-beta][v0.1.6-beta] (2019-05-21)

- Added JUnit tag that allows to use arrays in parametized tests
- Fixed documentation generation
- Fixed docs missing headers
- Fixed `NullPointerException` when using the default `ObservedValue` constructor
- Increased code coverage
- Improved docs

## [0.1.5-beta.1][v0.1.5-beta.1] (2019-05-10)

- Added [SonarCloud](https://sonarcloud.io/dashboard?id=com.fermod%3Aevent-dispatcher) to the repository
- Added Travis CI automatic jar release
- Added more project information to `pom.xml`
- Removed the symbolic link and changed Doxygen docs output path, with the one pointed by the symbolic link

## [0.1.5-beta][v0.1.5-beta] (2019-05-05)

- Added issue templates:
  - `bug_report.md`
  - `feature_request.md`
  - `pull_request_template.md`
- Added documentation files:
  - `CHANGELOG.md`
  - `CONTRIBUTING.md`
  - `CODE_OF_CONDUCT.md`
- Added [Travis CI](https://travis-ci.org/FerMod/EventDispatcher) support
- Added `.gitattributes` and added `.editorconfig` rule to handle scripts line endigs in Unix style
- Moved scripts to a new folder called `scripts`
- Improved and fixed scripts
- Improved `README.md` file

## [0.1.4-beta.1][v0.1.4-beta.1] (2019-04-16)

- Automatized docs generation
- Improved bash scripts
- Cleaned up `pom.xml`

## [0.1.4-beta][v0.1.4-beta] (2019-04-15)

- Changed Maven site output folder
- Fixed typos

## [0.1.3-beta][v0.1.3-beta] (2019-04-05)

- Added script to update Maven version
- Added missing documentation
- Added random string generator class
- Renamed package

## [0.1.2-beta.2][v0.1.2-beta.2] (2019-04-03)

- Improved docs generation
- Cleaned up generated directory tree in `README.md` file

## [0.1.2-beta][v0.1.2-beta] (019-04-02)

- Added equality tests
- Fixed typos

## [0.1.1-beta][v0.1.2-beta] (2019-03-28)

- Migrated from Log4j to Log4j2

## [0.1.0-beta][v0.1.0-beta] (2019-03-26)

- First release

[unreleased]: https://github.com/FerMod/EventDispatcher/compare/v0.1.6-beta.1...HEAD
[v0.1.6-beta.1]: https://github.com/FerMod/EventDispatcher/compare/v0.1.6-beta...v0.1.6-beta.1
[v0.1.6-beta]: https://github.com/FerMod/EventDispatcher/compare/v0.1.5-beta.1...v0.1.6-beta
[v0.1.5-beta.1]: https://github.com/FerMod/EventDispatcher/compare/v0.1.5-beta...v0.1.5-beta.1
[v0.1.5-beta]: https://github.com/FerMod/EventDispatcher/compare/v0.1.4-beta.1...v0.1.5-beta
[v0.1.4-beta.1]: https://github.com/FerMod/EventDispatcher/compare/v0.1.4-beta...v0.1.4-beta.1
[v0.1.4-beta]: https://github.com/FerMod/EventDispatcher/compare/v0.1.3-beta...v0.1.4-beta
[v0.1.3-beta]: https://github.com/FerMod/EventDispatcher/compare/v0.1.2-beta.2...v0.1.3-beta
[v0.1.2-beta.2]: https://github.com/FerMod/EventDispatcher/compare/v0.1.2-beta...v0.1.2-beta.2
[v0.1.2-beta]: https://github.com/FerMod/EventDispatcher/compare/v0.1.1-beta...v0.1.2-beta
[v0.1.1-beta]: https://github.com/FerMod/EventDispatcher/compare/v0.1.0-beta...v0.1.1-beta
[v0.1.0-beta]: https://github.com/FerMod/EventDispatcher/releases/tag/v0.1.0-beta


navtree.js
var NAVTREE =
[
  [ "Event Dispatcher", "index.html", [
    [ "Project Maven Site", "site/index.html", null ],
    [ "Project Tree", "tree.html", null ],
    [ "Packages", null, [
      [ "Packages", "namespaces.html", "namespaces" ]
    ] ]
  ] ]
];

var NAVTREEINDEX =
[
"index.html"
];

var SYNCONMSG = 'click to disable panel synchronisation';
var SYNCOFFMSG = 'click to enable panel synchronisation';
var SYNCONMSG = 'click to disable panel synchronisation';
var SYNCOFFMSG = 'click to enable panel synchronisation';
var navTreeSubIndices = new Array();

function getData(varName)
{
  var i = varName.lastIndexOf('/');
  var n = i>=0 ? varName.substring(i+1) : varName;
  return eval(n.replace(/\-/g,'_'));
}

function stripPath(uri)
{
  return uri.substring(uri.lastIndexOf('/')+1);
}

function stripPath2(uri)
{
  var i = uri.lastIndexOf('/');
  var s = uri.substring(i+1);
  var m = uri.substring(0,i+1).match(/\/d\w\/d\w\w\/$/);
  return m ? uri.substring(i-6) : s;
}

function localStorageSupported()
{
  try {
    return 'localStorage' in window && window['localStorage'] !== null && window.localStorage.getItem;
  }
  catch(e) {
    return false;
  }
}


function storeLink(link)
{
  if (!$("#nav-sync").hasClass('sync') && localStorageSupported()) {
      window.localStorage.setItem('navpath',link);
  }
}

function deleteLink()
{
  if (localStorageSupported()) {
    window.localStorage.setItem('navpath','');
  } 
}

function cachedLink()
{
  if (localStorageSupported()) {
    return window.localStorage.getItem('navpath');
  } else {
    return '';
  }
}

function getScript(scriptName,func,show)
{
  var head = document.getElementsByTagName("head")[0]; 
  var script = document.createElement('script');
  script.id = scriptName;
  script.type = 'text/javascript';
  script.onload = func; 
  script.src = scriptName+'.js'; 
  if ($.browser.msie && $.browser.version<=8) { 
    // script.onload does not work with older versions of IE
    script.onreadystatechange = function() {
      if (script.readyState=='complete' || script.readyState=='loaded') { 
        func(); if (show) showRoot(); 
      }
    }
  }
  head.appendChild(script); 
}

function createIndent(o,domNode,node,level)
{
  var level=-1;
  var n = node;
  while (n.parentNode) { level++; n=n.parentNode; }
  if (node.childrenData) {
    var imgNode = document.createElement("img");
    imgNode.style.paddingLeft=(16*level).toString()+'px';
    imgNode.width  = 16;
    imgNode.height = 22;
    imgNode.border = 0;
    node.plus_img = imgNode;
    node.expandToggle = document.createElement("a");
    node.expandToggle.href = "javascript:void(0)";
    node.expandToggle.onclick = function() {
      if (node.expanded) {
        $(node.getChildrenUL()).slideUp("fast");
        node.plus_img.src = node.relpath+"ftv2pnode.png";
        node.expanded = false;
      } else {
        expandNode(o, node, false, false);
      }
    }
    node.expandToggle.appendChild(imgNode);
    domNode.appendChild(node.expandToggle);
    imgNode.src = node.relpath+"ftv2pnode.png";
  } else {
    var span = document.createElement("span");
    span.style.display = 'inline-block';
    span.style.width   = 16*(level+1)+'px';
    span.style.height  = '22px';
    span.innerHTML = '&#160;';
    domNode.appendChild(span);
  } 
}

var animationInProgress = false;

function gotoAnchor(anchor,aname,updateLocation)
{
  var pos, docContent = $('#doc-content');
  if (anchor.parent().attr('class')=='memItemLeft' ||
      anchor.parent().attr('class')=='fieldtype' ||
      anchor.parent().is(':header')) 
  {
    pos = anchor.parent().position().top;
  } else if (anchor.position()) {
    pos = anchor.position().top;
  }
  if (pos) {
    var dist = Math.abs(Math.min(
               pos-docContent.offset().top,
               docContent[0].scrollHeight-
               docContent.height()-docContent.scrollTop()));
    animationInProgress=true;
    docContent.animate({
      scrollTop: pos + docContent.scrollTop() - docContent.offset().top
    },Math.max(50,Math.min(500,dist)),function(){
      if (updateLocation) window.location.href=aname;
      animationInProgress=false;
    });
  }
}

function newNode(o, po, text, link, childrenData, lastNode)
{
  var node = new Object();
  node.children = Array();
  node.childrenData = childrenData;
  node.depth = po.depth + 1;
  node.relpath = po.relpath;
  node.isLast = lastNode;

  node.li = document.createElement("li");
  po.getChildrenUL().appendChild(node.li);
  node.parentNode = po;

  node.itemDiv = document.createElement("div");
  node.itemDiv.className = "item";

  node.labelSpan = document.createElement("span");
  node.labelSpan.className = "label";

  createIndent(o,node.itemDiv,node,0);
  node.itemDiv.appendChild(node.labelSpan);
  node.li.appendChild(node.itemDiv);

  var a = document.createElement("a");
  node.labelSpan.appendChild(a);
  node.label = document.createTextNode(text);
  node.expanded = false;
  a.appendChild(node.label);
  if (link) {
    var url;
    if (link.substring(0,1)=='^') {
      url = link.substring(1);
      link = url;
    } else {
      url = node.relpath+link;
    }
    a.className = stripPath(link.replace('#',':'));
    if (link.indexOf('#')!=-1) {
      var aname = '#'+link.split('#')[1];
      var srcPage = stripPath($(location).attr('pathname'));
      var targetPage = stripPath(link.split('#')[0]);
      a.href = srcPage!=targetPage ? url : "javascript:void(0)"; 
      a.onclick = function(){
        storeLink(link);
        if (!$(a).parent().parent().hasClass('selected'))
        {
          $('.item').removeClass('selected');
          $('.item').removeAttr('id');
          $(a).parent().parent().addClass('selected');
          $(a).parent().parent().attr('id','selected');
        }
        var anchor = $(aname);
        gotoAnchor(anchor,aname,true);
      };
    } else {
      a.href = url;
      a.onclick = function() { storeLink(link); }
    }
  } else {
    if (childrenData != null) 
    {
      a.className = "nolink";
      a.href = "javascript:void(0)";
      a.onclick = node.expandToggle.onclick;
    }
  }

  node.childrenUL = null;
  node.getChildrenUL = function() {
    if (!node.childrenUL) {
      node.childrenUL = document.createElement("ul");
      node.childrenUL.className = "children_ul";
      node.childrenUL.style.display = "none";
      node.li.appendChild(node.childrenUL);
    }
    return node.childrenUL;
  };

  return node;
}

function showRoot()
{
  var headerHeight = $("#top").height();
  var footerHeight = $("#nav-path").height();
  var windowHeight = $(window).height() - headerHeight - footerHeight;
  (function (){ // retry until we can scroll to the selected item
    try {
      var navtree=$('#nav-tree');
      navtree.scrollTo('#selected',0,{offset:-windowHeight/2});
    } catch (err) {
      setTimeout(arguments.callee, 0);
    }
  })();
}

function expandNode(o, node, imm, showRoot)
{
  if (node.childrenData && !node.expanded) {
    if (typeof(node.childrenData)==='string') {
      var varName    = node.childrenData;
      getScript(node.relpath+varName,function(){
        node.childrenData = getData(varName);
        expandNode(o, node, imm, showRoot);
      }, showRoot);
    } else {
      if (!node.childrenVisited) {
        getNode(o, node);
      } if (imm || ($.browser.msie && $.browser.version>8)) { 
        // somehow slideDown jumps to the start of tree for IE9 :-(
        $(node.getChildrenUL()).show();
      } else {
        $(node.getChildrenUL()).slideDown("fast");
      }
      if (node.isLast) {
        node.plus_img.src = node.relpath+"ftv2mlastnode.png";
      } else {
        node.plus_img.src = node.relpath+"ftv2mnode.png";
      }
      node.expanded = true;
    }
  }
}

function glowEffect(n,duration)
{
  n.addClass('glow').delay(duration).queue(function(next){
    $(this).removeClass('glow');next();
  });
}

function highlightAnchor()
{
  var aname = $(location).attr('hash');
  var anchor = $(aname);
  if (anchor.parent().attr('class')=='memItemLeft'){
    var rows = $('.memberdecls tr[class$="'+
               window.location.hash.substring(1)+'"]');
    glowEffect(rows.children(),300); // member without details
  } else if (anchor.parents().slice(2).prop('tagName')=='TR') {
    glowEffect(anchor.parents('div.memitem'),1000); // enum value
  } else if (anchor.parent().attr('class')=='fieldtype'){
    glowEffect(anchor.parent().parent(),1000); // struct field
  } else if (anchor.parent().is(":header")) {
    glowEffect(anchor.parent(),1000); // section header
  } else {
    glowEffect(anchor.next(),1000); // normal member
  }
  gotoAnchor(anchor,aname,false);
}

function selectAndHighlight(hash,n)
{
  var a;
  if (hash) {
    var link=stripPath($(location).attr('pathname'))+':'+hash.substring(1);
    a=$('.item a[class$="'+link+'"]');
  }
  if (a && a.length) {
    a.parent().parent().addClass('selected');
    a.parent().parent().attr('id','selected');
    highlightAnchor();
  } else if (n) {
    $(n.itemDiv).addClass('selected');
    $(n.itemDiv).attr('id','selected');
  }
  if ($('#nav-tree-contents .item:first').hasClass('selected')) {
    $('#nav-sync').css('top','30px');
  } else {
    $('#nav-sync').css('top','5px');
  }
  showRoot();
}

function showNode(o, node, index, hash)
{
  if (node && node.childrenData) {
    if (typeof(node.childrenData)==='string') {
      var varName    = node.childrenData;
      getScript(node.relpath+varName,function(){
        node.childrenData = getData(varName);
        showNode(o,node,index,hash);
      },true);
    } else {
      if (!node.childrenVisited) {
        getNode(o, node);
      }
      $(node.getChildrenUL()).css({'display':'block'});
      if (node.isLast) {
        node.plus_img.src = node.relpath+"ftv2mlastnode.png";
      } else {
        node.plus_img.src = node.relpath+"ftv2mnode.png";
      }
      node.expanded = true;
      var n = node.children[o.breadcrumbs[index]];
      if (index+1<o.breadcrumbs.length) {
        showNode(o,n,index+1,hash);
      } else {
        if (typeof(n.childrenData)==='string') {
          var varName = n.childrenData;
          getScript(n.relpath+varName,function(){
            n.childrenData = getData(varName);
            node.expanded=false;
            showNode(o,node,index,hash); // retry with child node expanded
          },true);
        } else {
          var rootBase = stripPath(o.toroot.replace(/\..+$/, ''));
          if (rootBase=="index" || rootBase=="pages" || rootBase=="search") {
            expandNode(o, n, true, true);
          }
          selectAndHighlight(hash,n);
        }
      }
    }
  } else {
    selectAndHighlight(hash);
  }
}

function removeToInsertLater(element) {
  var parentNode = element.parentNode;
  var nextSibling = element.nextSibling;
  parentNode.removeChild(element);
  return function() {
    if (nextSibling) {
      parentNode.insertBefore(element, nextSibling);
    } else {
      parentNode.appendChild(element);
    }
  };
}

function getNode(o, po)
{
  var insertFunction = removeToInsertLater(po.li);
  po.childrenVisited = true;
  var l = po.childrenData.length-1;
  for (var i in po.childrenData) {
    var nodeData = po.childrenData[i];
    po.children[i] = newNode(o, po, nodeData[0], nodeData[1], nodeData[2],
      i==l);
  }
  insertFunction();
}

function gotoNode(o,subIndex,root,hash,relpath)
{
  var nti = navTreeSubIndices[subIndex][root+hash];
  o.breadcrumbs = $.extend(true, [], nti ? nti : navTreeSubIndices[subIndex][root]);
  if (!o.breadcrumbs && root!=NAVTREE[0][1]) { // fallback: show index
    navTo(o,NAVTREE[0][1],"",relpath);
    $('.item').removeClass('selected');
    $('.item').removeAttr('id');
  }
  if (o.breadcrumbs) {
    o.breadcrumbs.unshift(0); // add 0 for root node
    showNode(o, o.node, 0, hash);
  }
}

function navTo(o,root,hash,relpath)
{
  var link = cachedLink();
  if (link) {
    var parts = link.split('#');
    root = parts[0];
    if (parts.length>1) hash = '#'+parts[1];
    else hash='';
  }
  if (hash.match(/^#l\d+$/)) {
    var anchor=$('a[name='+hash.substring(1)+']');
    glowEffect(anchor.parent(),1000); // line number
    hash=''; // strip line number anchors
    //root=root.replace(/_source\./,'.'); // source link to doc link
  }
  var url=root+hash;
  var i=-1;
  while (NAVTREEINDEX[i+1]<=url) i++;
  if (i==-1) { i=0; root=NAVTREE[0][1]; } // fallback: show index
  if (navTreeSubIndices[i]) {
    gotoNode(o,i,root,hash,relpath)
  } else {
    getScript(relpath+'navtreeindex'+i,function(){
      navTreeSubIndices[i] = eval('NAVTREEINDEX'+i);
      if (navTreeSubIndices[i]) {
        gotoNode(o,i,root,hash,relpath);
      }
    },true);
  }
}

function showSyncOff(n,relpath)
{
    n.html('<img src="'+relpath+'sync_off.png" title="'+SYNCOFFMSG+'"/>');
}

function showSyncOn(n,relpath)
{
    n.html('<img src="'+relpath+'sync_on.png" title="'+SYNCONMSG+'"/>');
}

function toggleSyncButton(relpath)
{
  var navSync = $('#nav-sync');
  if (navSync.hasClass('sync')) {
    navSync.removeClass('sync');
    showSyncOff(navSync,relpath);
    storeLink(stripPath2($(location).attr('pathname'))+$(location).attr('hash'));
  } else {
    navSync.addClass('sync');
    showSyncOn(navSync,relpath);
    deleteLink();
  }
}

function initNavTree(toroot,relpath)
{
  var o = new Object();
  o.toroot = toroot;
  o.node = new Object();
  o.node.li = document.getElementById("nav-tree-contents");
  o.node.childrenData = NAVTREE;
  o.node.children = new Array();
  o.node.childrenUL = document.createElement("ul");
  o.node.getChildrenUL = function() { return o.node.childrenUL; };
  o.node.li.appendChild(o.node.childrenUL);
  o.node.depth = 0;
  o.node.relpath = relpath;
  o.node.expanded = false;
  o.node.isLast = true;
  o.node.plus_img = document.createElement("img");
  o.node.plus_img.src = relpath+"ftv2pnode.png";
  o.node.plus_img.width = 16;
  o.node.plus_img.height = 22;

  if (localStorageSupported()) {
    var navSync = $('#nav-sync');
    if (cachedLink()) {
      showSyncOff(navSync,relpath);
      navSync.removeClass('sync');
    } else {
      showSyncOn(navSync,relpath);
    }
    navSync.click(function(){ toggleSyncButton(relpath); });
  }

  $(window).load(function(){
    navTo(o,toroot,window.location.hash,relpath);
    showRoot();
  });

  $(window).bind('hashchange', function(){
     if (window.location.hash && window.location.hash.length>1){
       var a;
       if ($(location).attr('hash')){
         var clslink=stripPath($(location).attr('pathname'))+':'+
                               $(location).attr('hash').substring(1);
         a=$('.item a[class$="'+clslink+'"]');
       }
       if (a==null || !$(a).parent().parent().hasClass('selected')){
         $('.item').removeClass('selected');
         $('.item').removeAttr('id');
       }
       var link=stripPath2($(location).attr('pathname'));
       navTo(o,link,$(location).attr('hash'),relpath);
     } else if (!animationInProgress) {
       $('#doc-content').scrollTop(0);
       $('.item').removeClass('selected');
       $('.item').removeAttr('id');
       navTo(o,toroot,window.location.hash,relpath);
     }
  })
}

